lib/component.js

```diff
@@ -805,6 +817,113 @@ class Component extends WebComponent {
     return this._attrs;
   }
 
+  _syncParams() {
+    // maintain local validated map where all schema keys are defined
+    this._paramTypes = {};
+    const paramTypes = this.constructor.paramTypes || {};
+
+    // vuejs like this.data.$data to reference "real" value of params
+    Object.defineProperty(this.params, `$params`, {
+      enumerable: false,
+      value: {},
+    });
+
+    for (const param of Object.keys(paramTypes)) {
+      // convert type shorthand to object
+      let paramType = paramTypes[param];
+      if (PARAM_TYPE_DEFAULTS.has(paramType)) {
+        paramType = {type: paramType};
+      }
+
+      // Ensure param type is valid
+      const type = paramType.type;
+      if (!PARAM_TYPE_DEFAULTS.has(type)) {
+        const typeString = typeof type === `function` ? type.name : String(type);
+        throw new Error(
+          `Invalid type: ${typeString} for attr: ${param} in paramType. ` +
+            `Only (${Array.from(PARAM_TYPE_DEFAULTS.keys())
+              .map((v) => `'${v.name}'`)
+              .join(` | `)}) is valid.`,
+        );
+      }
+
+      if (paramType.default && paramType.required) {
+        throw new Error(`${this}: param '${param}' cannot have both required and default`);
+      }
+
+      if ([Object, Array, Map, Set].includes(type) && !paramType.shouldUpdate) {
+        throw new Error(
+          `Object like params '${param}' on ${this.constructor.name} must have 'shouldUpdate' function passed in`,
+        );
+      }
+
+      const paramSchemaObj = {
+        type,
+        default: paramType.default ? paramType.default : PARAM_TYPE_DEFAULTS.get(type),
+        required: paramType.required ? paramType.required : false,
+        // by default use === for comparision
+        shouldUpdate: paramType.shouldUpdate || ((oldVal, newVal) => oldVal !== newVal),
+      };
+
+      // set default value for the params
+      this.params.$params[param] = paramSchemaObj.default;
+
+      this._paramTypes[param] = paramSchemaObj;
+    }
+
+    const paramDescription = Object.keys(paramTypes)
+      .map((param) => [
+        param,
+        {
+          get() {
+            return this.$params[param];
+          },
+        },
+      ])
+      .reduce((acc, cur) => {
+        acc[cur[0]] = cur[1];
+        return acc;
+      }, {});
+    Object.defineProperties(this.params, paramDescription);
+    Object.freeze(this.params);
+    return this._paramTypes;
+  }
+
+  _enqueueParamUpdate(param, value) {
+    this._count = 0;
+    if (this._queuedParamUpdates) {
+      this._queuedParamUpdates[param] = value;
+      return;
+    }
+    this._queuedParamUpdates = {[param]: value};
+    queueMicrotask(() => {
+      const options = {
```

Comment:
@themacguffinman updated a bit of update mechanism for here. to make sure `this.params` points to the old value, I need to run hooks and update manually here.

Reply:
It's not clear to me that `this._enqueueParamUpdate()` should exist as a separate method, nor is there a need to use `queueMicrotask()`. Just run the hooks inside `_setParams()`.

Reply:
yeah now that i think twice, seem it's not possible to have 2 passes of `domPatcher.patch` to update params, so a simple method could do the trick